pulumi convert doesn't respect language specific overrides for properties in schema for charp

### What happened?

Running pulumi convert on a resource with properties with language specific overrides (I have really only tested this with C#)
 seems to not be honored.

e.g. with the following Pulumi.yaml contents (copy the contents to a Pulumi.yaml):
https://github.com/pulumi/pulumi-kubernetes/blob/f93f9b6c75303e3c600b2d1f5f09a78dd6ab9278/provider/pkg/gen/examples/upstream/yaml/deployment.yaml

results in the following:
```
pulumi convert --language csharp --out /tmp/example-dotnet
Installing dependencies...

MSBuild version 17.3.0+92e077650 for .NET
  Determining projects to restore...
  Restored /private/tmp/example-dotnet/yaml-example.csproj (in 872 ms).
/private/tmp/example-dotnet/Program.cs(47,37): error CS0117: 'ContainerPortArgs' does not contain a definition for 'ContainerPort' [/private/tmp/example-dotnet/yaml-example.csproj]

Build FAILED.

/private/tmp/example-dotnet/Program.cs(47,37): error CS0117: 'ContainerPortArgs' does not contain a definition for 'ContainerPort' [/private/tmp/example-dotnet/yaml-example.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:03.51
error: installing dependencies failed; rerun manually to try again, then run `pulumi up` to perform an initial deployment: `dotnet build` failed to install dependencies: exit status 1
```

In particular the containerport field has the following definition in the schema:
```
"kubernetes:core/v1:ContainerPort": {
            "description": "ContainerPort represents a network port in a single container.",
            "properties": {
                "containerPort": {
                    "type": "integer",
                    "description": "Number of port to expose on the pod's IP address. This must be a valid port number, 0 < x < 65536.",
                    "language": {
                        "csharp": {
                            "name": "ContainerPortValue"
                        }
                    }
                },
...
```

### Steps to reproduce

`pulumi convert --language csharp --out /tmp/example-dotnet` with contents from https://github.com/pulumi/pulumi-kubernetes/blob/f93f9b6c75303e3c600b2d1f5f09a78dd6ab9278/provider/pkg/gen/examples/upstream/yaml/deployment.yaml


### Expected Behavior

Successful conversion to csharp

### Actual Behavior

```
/private/tmp/example-dotnet/Program.cs(47,37): error CS0117: 'ContainerPortArgs' does not contain a definition for 'ContainerPort' [/private/tmp/example-dotnet/yaml-example.csproj]

Build FAILED.
```

### Output of `pulumi about`

```
CLI
Version      3.39.3
Go Version   go1.19.1
Go Compiler  gc

Plugins
NAME        VERSION
kubernetes  unknown
yaml        unknown

Host
OS       darwin
Version  11.6.8
Arch     x86_64

This project is written in yaml
```

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/changelog/pending/20221009--programgen-dotnet--pcl-annotate-resource-inputs-apply-property-overrides.yaml
+++ b/changelog/pending/20221009--programgen-dotnet--pcl-annotate-resource-inputs-apply-property-overrides.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: programgen/dotnet
+  description: Annotate deeply nested objects with their schema types and apply property name overrides

--- a/pkg/codegen/dotnet/gen_program.go
+++ b/pkg/codegen/dotnet/gen_program.go
@@ -653,6 +653,8 @@ func (g *generator) genResource(w io.Writer, r *pcl.Resource) {
 		}
 	}
 
+	pcl.AnnotateResourceInputs(r)
+
 	name := r.LogicalName()
 	variableName := makeValidIdentifier(r.Name())
 	argsName := g.resourceArgsTypeName(r)

--- a/pkg/codegen/dotnet/gen_program_expressions.go
+++ b/pkg/codegen/dotnet/gen_program_expressions.go
@@ -663,6 +663,42 @@ func (g *generator) genObjectConsExpression(w io.Writer, expr *model.ObjectConsE
 	g.genObjectConsExpressionWithTypeName(w, expr, destTypeName, false)
 }
 
+func propertyNameOverrides(exprType model.Type) map[string]string {
+	overrides := make(map[string]string)
+	schemaType, ok := pcl.GetSchemaForType(exprType)
+	if !ok {
+		return overrides
+	}
+
+	switch arg := schemaType.(type) {
+	case *schema.ObjectType:
+		for _, property := range arg.Properties {
+			foundOverride := false
+			if csharp, ok := property.Language["csharp"]; ok {
+				if options, ok := csharp.(CSharpPropertyInfo); ok {
+					overrides[property.Name] = options.Name
+					foundOverride = true
+				}
+			}
+
+			if !foundOverride {
+				overrides[property.Name] = property.Name
+			}
+		}
+	}
+
+	return overrides
+}
+
+func resolvePropertyName(property string, overrides map[string]string) string {
+	foundOverride, ok := overrides[property]
+	if ok {
+		return propertyName(foundOverride)
+	}
+
+	return propertyName(property)
+}
+
 func (g *generator) genObjectConsExpressionWithTypeName(
 	w io.Writer, expr *model.ObjectConsExpression, destTypeName string, implicitTypeName bool) {
 
@@ -678,12 +714,14 @@ func (g *generator) genObjectConsExpressionWithTypeName(
 			g.Fgenf(w, "new %s", typeName)
 		}
 
+		propertyNames := propertyNameOverrides(expr.Type())
 		g.Fgenf(w, "\n%s{\n", g.Indent)
 		g.Indented(func() {
 			for _, item := range expr.Items {
 				g.Fgenf(w, "%s", g.Indent)
 				lit := item.Key.(*model.LiteralValueExpression)
-				g.Fprint(w, propertyName(lit.Value.AsString()))
+				propertyKey := lit.Value.AsString()
+				g.Fprint(w, resolvePropertyName(propertyKey, propertyNames))
 				g.Fgenf(w, " = %.v,\n", item.Value)
 			}
 		})

--- a/pkg/codegen/hcl2/model/expression.go
+++ b/pkg/codegen/hcl2/model/expression.go
@@ -1456,6 +1456,10 @@ func (x *ObjectConsExpression) Type() Type {
 	return x.exprType
 }
 
+func (x *ObjectConsExpression) WithType(updateType func(Type) *ObjectConsExpression) *ObjectConsExpression {
+	return updateType(x.exprType)
+}
+
 func (x *ObjectConsExpression) Typecheck(typecheckOperands bool) hcl.Diagnostics {
 	var diagnostics hcl.Diagnostics
 

--- a/pkg/codegen/pcl/binder_resource.go
+++ b/pkg/codegen/pcl/binder_resource.go
@@ -42,6 +42,82 @@ func (b *binder) bindResource(node *Resource) hcl.Diagnostics {
 	return diagnostics
 }
 
+func annotateAttributeValue(expr model.Expression, attributeType schema.Type) model.Expression {
+	if optionalType, ok := attributeType.(*schema.OptionalType); ok {
+		return annotateAttributeValue(expr, optionalType.ElementType)
+	}
+
+	switch attrValue := expr.(type) {
+	case *model.ObjectConsExpression:
+		if schemaObjectType, ok := attributeType.(*schema.ObjectType); ok {
+			schemaProperties := make(map[string]schema.Type)
+			for _, schemaProperty := range schemaObjectType.Properties {
+				schemaProperties[schemaProperty.Name] = schemaProperty.Type
+			}
+
+			for _, item := range attrValue.Items {
+				keyLiteral, isLit := item.Key.(*model.LiteralValueExpression)
+				if isLit {
+					correspondingSchemaType, ok := schemaProperties[keyLiteral.Value.AsString()]
+					if ok {
+						item.Value = annotateAttributeValue(item.Value, correspondingSchemaType)
+					}
+				}
+			}
+			return attrValue.WithType(func(attrValueType model.Type) *model.ObjectConsExpression {
+				annotateObjectProperties(attrValueType, attributeType)
+				return attrValue
+			})
+		}
+
+		return attrValue
+
+	case *model.TupleConsExpression:
+		if schemaArrayType, ok := attributeType.(*schema.ArrayType); ok {
+			elementType := schemaArrayType.ElementType
+			for _, arrayExpr := range attrValue.Expressions {
+				annotateAttributeValue(arrayExpr, elementType)
+			}
+		}
+
+		return attrValue
+	case *model.FunctionCallExpression:
+		if attrValue.Name == IntrinsicConvert {
+			converterArg := attrValue.Args[0]
+			annotateAttributeValue(converterArg, attributeType)
+		}
+
+		return attrValue
+	default:
+		return expr
+	}
+}
+
+func AnnotateAttributeValue(expr model.Expression, attributeType schema.Type) model.Expression {
+	return annotateAttributeValue(expr, attributeType)
+}
+
+func AnnotateResourceInputs(node *Resource) {
+	resourceProperties := make(map[string]*schema.Property)
+	for _, property := range node.Schema.Properties {
+		resourceProperties[property.Name] = property
+	}
+
+	// add type annotations to the attributes
+	// and their nested objects
+	for index := range node.Inputs {
+		attr := node.Inputs[index]
+		if property, ok := resourceProperties[attr.Name]; ok {
+			node.Inputs[index] = &model.Attribute{
+				Tokens: attr.Tokens,
+				Name:   attr.Name,
+				Syntax: attr.Syntax,
+				Value:  AnnotateAttributeValue(attr.Value, property.Type),
+			}
+		}
+	}
+}
+
 // bindResourceTypes binds the input and output types for a resource.
 func (b *binder) bindResourceTypes(node *Resource) hcl.Diagnostics {
 	// Set the input and output types to dynamic by default.
@@ -110,6 +186,7 @@ func (b *binder) bindResourceTypes(node *Resource) hcl.Diagnostics {
 	outputType := model.NewObjectType(outputProperties, &schema.ObjectType{Properties: properties})
 
 	node.InputType, node.OutputType = inputType, outputType
+
 	return diagnostics
 }
 

--- a/pkg/codegen/testing/test/testdata/kubernetes-pod-pp/dotnet/kubernetes-pod.cs
+++ b/pkg/codegen/testing/test/testdata/kubernetes-pod-pp/dotnet/kubernetes-pod.cs
@@ -21,6 +21,13 @@ return await Deployment.RunAsync(() =>
                 {
                     Name = "nginx",
                     Image = "nginx:1.14-alpine",
+                    Ports = new[]
+                    {
+                        new Kubernetes.Types.Inputs.Core.V1.ContainerPortArgs
+                        {
+                            ContainerPortValue = 80,
+                        },
+                    },
                     Resources = new Kubernetes.Types.Inputs.Core.V1.ResourceRequirementsArgs
                     {
                         Limits = 

--- a/pkg/codegen/testing/test/testdata/kubernetes-pod-pp/go/kubernetes-pod.go
+++ b/pkg/codegen/testing/test/testdata/kubernetes-pod-pp/go/kubernetes-pod.go
@@ -20,6 +20,11 @@ func main() {
 					&corev1.ContainerArgs{
 						Name:  pulumi.String("nginx"),
 						Image: pulumi.String("nginx:1.14-alpine"),
+						Ports: corev1.ContainerPortArray{
+							&corev1.ContainerPortArgs{
+								ContainerPort: pulumi.Int(80),
+							},
+						},
 						Resources: &corev1.ResourceRequirementsArgs{
 							Limits: pulumi.StringMap{
 								"memory": pulumi.String("20Mi"),

--- a/pkg/codegen/testing/test/testdata/kubernetes-pod-pp/kubernetes-pod.pp
+++ b/pkg/codegen/testing/test/testdata/kubernetes-pod-pp/kubernetes-pod.pp
@@ -10,6 +10,7 @@ resource bar "kubernetes:core/v1:Pod" {
             {
                 name = "nginx"
                 image = "nginx:1.14-alpine"
+                ports = [{ containerPort = 80 }]
                 resources = {
                     limits = {
                         memory = "20Mi"

--- a/pkg/codegen/testing/test/testdata/kubernetes-pod-pp/nodejs/kubernetes-pod.ts
+++ b/pkg/codegen/testing/test/testdata/kubernetes-pod-pp/nodejs/kubernetes-pod.ts
@@ -12,6 +12,9 @@ const bar = new kubernetes.core.v1.Pod("bar", {
         containers: [{
             name: "nginx",
             image: "nginx:1.14-alpine",
+            ports: [{
+                containerPort: 80,
+            }],
             resources: {
                 limits: {
                     memory: "20Mi",

--- a/pkg/codegen/testing/test/testdata/kubernetes-pod-pp/python/kubernetes-pod.py
+++ b/pkg/codegen/testing/test/testdata/kubernetes-pod-pp/python/kubernetes-pod.py
@@ -12,6 +12,9 @@ bar = kubernetes.core.v1.Pod("bar",
         containers=[kubernetes.core.v1.ContainerArgs(
             name="nginx",
             image="nginx:1.14-alpine",
+            ports=[kubernetes.core.v1.ContainerPortArgs(
+                container_port=80,
+            )],
             resources=kubernetes.core.v1.ResourceRequirementsArgs(
                 limits={
                     "memory": "20Mi",

